Support the UAST dataset in `pga` tool

r0mainK opened the pull request:
For more info on context see [this ml issue](https://github.com/src-d/ml-backlog/issues/97) and  [this infra issue](https://github.com/src-d/infrastructure/issues/1170).

This PR only deals with the first dataset, ie the UASTs extracted from the HEAD of PGAv2, and stored as parquet. It does not contain yet any documentation, apart from in side the code - and I did not put much effort into the comments, as I wasn't sure the logic would be kept. Anyway here's a run down of the changes:

**commit 1**: typos

**commit 2 and 3**: 

Add `uast_dataset.go` and `siva_dataset.go`. Basically, since both datasets are comprised of individual files I wanted to keep the same commands for listing and downloading, so I had to abstract all the logic from `IndexToCSV`, RepositoryFromCSV and `ToCSV`, as well as the schemas. In order to do that I:
- transformed `Repository` in an interface with `ToCSV` as it's method
- added a Dataset interface with `IndexFromCSV` as it's method
- renamed `RepositoryFromCSV` to be dataset specific, and handled by dataset specific index - as we don't need to deal with legacy indexes for the new dataset.
- I then added schemas for both datasets (had to rename most variables of the Siva dataset)
- added getters for the abstract `Repository`'s `URL`, `Languages` and `Siva`/`Parquet` `Filenames`, used in other places.

**commit 4**  

Following the previous changes, I basically removed all code I move to `siva_dataset`, added parsers/formatter for `floats` with 2 point precision as well as a `csvColumn` field, and created a mapping for datasets. Also modified the types as `Repository` is now an interface, and (I think) can't be pointed to the same.

**commit 5**

Now using getters in `filters.go`, and updated typing.

**commit 6**

Added a handler for the `datasetName` arg now used in `get`/`list`.

**commit 7**

Updated commands to the new structure, more of the same: getters, typing, now using `Dataset` and `Dataset` handler.


**Note:** I was thinking of adding a parquet command to dump listing of individual parquet files, as well as some additional filters to be used only by the `uast` dataset, related to the extraction rate. What do you think ?

vmarkovtsev commented on PublicGitArchive/pga/cmd/cache.go:
```diff
@@ -125,13 +125,13 @@ func matchHash(dest, source FileSystem, name string) bool {
 }
 
-func getIndex(ctx context.Context) (io.ReadCloser, error) {
+func getIndex(ctx context.Context, datasetName string) (io.ReadCloser, error) {
```
Let's have an enum here

```go
type DatasetKind int

const (
    DatasetSiva DatasetKind = iota
    DaatsetUast DatasetKind = iota
)
```

r0mainK replied:
Why ? I mean here we just need the name to join with the url/cache, and elsewhere we don't really need an index with the `Name` methode

vmarkovtsev replied:
ah ok

vmarkovtsev commented on PublicGitArchive/pga/cmd/cache.go:
```diff
@@ -11,7 +11,7 @@ import (
 )
 
-const indexURL = "http://pga.sourced.tech/csv"
+const indexURL = "http://pga.sourced.tech/csv/"
```
We shall always use `path.Join` to concat URLs (note: it is different from `filepath`)

r0mainK replied:
I did it at first but is snips on of the `/` in the url, returning `http:/pga.sourced.tech/csv/uast`

r0mainK replied:
oke doing like https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -3,202 +3,46 @@
-	return cp
+// Dataset provides abstraction for creating an Index from a CSV file.
+type Dataset interface {
+	IndexFromCSV(r io.Reader) (Index, error)
```
Add `Name() string` and refactor `handleDatasetArg`

r0mainK replied:
done

vmarkovtsev commented on PublicGitArchive/pga/cmd/get.go:
```diff
@@ -17,6 +17,8 @@ import (
 	pb "gopkg.in/cheggaaa/pb.v1"
 )
 
+const rootURL = "http://pga.sourced.tech/"
```
Wow I can't believe this was hardcoded before. Shame

r0mainK replied:
haha :p I guess nobody wanted to mess with Francesc's code :-1:

vmarkovtsev commented on PublicGitArchive/pga/cmd/root.go:
```diff
@@ -31,7 +31,7 @@ For more info, check http://pga.sourced.tech/`,
 
 		pgaVersion = pv
-		indexName = pv + ".csv.gz"
+		indexName = pv + ".index.csv.gz"
```
Is this compatible with the siva dataset?

r0mainK replied:
[Not yet](https://github.com/src-d/infrastructure/issues/1170#issuecomment-539444851) but it seemed better

vmarkovtsev commented on PublicGitArchive/pga/cmd/siva.go:
```diff
@@ -221,7 +221,7 @@ func dumpFiles(commit *object.Commit, destDir string) error {
 }
 
-// sivaCmd represents the list command
+// sivaCmd represents the siva command
```
I wrote a pretty bad comment :D

```suggestion
// sivaCmd represents the set of commands to work with the siva files: extract revisions, list, dump raw contents.
```

r0mainK replied:
:100:

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -3,202 +3,70 @@
+type errorBadHeaderLength struct {
+	expectedMin int
+	expectedMax int
+	length      int
```
```suggestion
type badHeaderLengthError struct {
    error
    expectedMin int
    expectedMax int
    length      int
```

r0mainK replied:
done

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -3,202 +3,70 @@
-	return cp
+// Dataset provides abstraction for creating an Index from a CSV file.
+type Dataset interface {
+	IndexFromCSV(r io.Reader) (Index, error)
```
```suggestion
	LoadIndex(r csv.Reader) (Index, error)
```

from `encoding/csv`

r0mainK replied:
done

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -3,202 +3,70 @@
-	r         *csv.Reader
-	withStars bool
-	withSize  bool
+type errorBadHeaderColumn struct {
```
```suggestion
type badHeaderColumnError struct {
    error
```

r0mainK replied:
done

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -3,202 +3,70 @@
-	return &csvIndex{cr, withStars, withSize}, nil
+// Index provides iteration over repositories.
+type Index interface {
+	Next() (Repository, error)
```
I am totally confused at this point. What is going on with Index, Repository and Dataset, what is their relationship?

It is more common to provide `ForEach(func(r Repository))` instead of `Next()`.

vmarkovtsev replied:
We should have crystal clear entities. We should also rip away "Repository" from everywhere because it confuses very much (Git? nope, gotcha).

Depending on how you describe the current logic, I can suggest the appropriate abstractions.

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -257,7 +126,42 @@ func (p *parser) intList(idx int) []int64 {
+	return vs
+}
+
+func (p *parser) float(idx int) float64 {
```
```suggestion
func (p *parser) ReadFloat(idx int) float64 {
```

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -257,7 +126,42 @@ func (p *parser) intList(idx int) []int64 {
+	return v
+}
+
+func (p *parser) floatList(idx int) []float64 {
```
```suggestion
func (p *parser) ReadListOfFloats(idx int) []float64 {
```

r0mainK commented:
@vmarkovtsev this is rebased and modified according to the [first part](https://github.com/src-d/datasets/pull/167).  In order to avoid some redundancy between `siva.go` and `uast.go` I replaced the `ForEach` method with a `ForEachRepository` function that takes the `dataset` as input. This allowed me to refactor `Uast/SivaRepositoryFromCSV` functions in a common `RepositoryFromCSV` interface method.

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -23,12 +24,41 @@ type Filter func(Repository) bool
 	Name() string
 	ReadHeader(r *csv.Reader) error
-	ForEach(ctx context.Context, r *csv.Reader, filter Filter, f func(r Repository) error) error
+	RepositoryFromCSV(cols []string) (repo Repository, err error)
```
```suggestion
	RepositoryFromTuple(cols []string) (repo Repository, err error)
```

r0mainK replied:
done in split

vmarkovtsev commented on PublicGitArchive/pga/pga/pga.go:
```diff
@@ -23,12 +24,41 @@ type Filter func(Repository) bool
 type Dataset interface {
 	Name() string
 	ReadHeader(r *csv.Reader) error
```
```suggestion
	ReadHeader(columnNames []string) error
```

This makes sense after you've added `pga.ForEachRepository`.

r0mainK replied:
:+1:

r0mainK commented:
Just added a couple commits: temporary files were not cleaned in case of command cancellation - furthermore the change I brought to stop updating the `Bar` in case of an issue with `pga get` also caused files to not be cleaned, in the case multiple files were being downloaded.

vmarkovtsev commented on PublicGitArchive/pga/pga/siva.go:
```diff
@@ -174,37 +149,31 @@ func (dataset *SivaDataset) ReadHeader(r *csv.Reader) error {
-			}
-		}
+// RepositoryFromCSV returns a SivaRepository from a slice of strings corresponding to it's CSV representation.
+func (dataset *SivaDataset) RepositoryFromCSV(cols []string) (repo Repository, err error) {
```
Can you please move this function to line 106 to minimize the diff size.

r0mainK commented:
@vmarkovtsev second split is done, all refactoring (4 fisrt commits) is in [this PR](https://github.com/src-d/datasets/pull/169)
